### PR TITLE
Added a ListCommandResource alongwith ListCommandResourceV2Provider

### PR DESCRIPTION
**Pull request by deepakaravindr**
and ListCommandResourceV3Provider. For v3, this resource will return the
v2 endpoint specified in index.json as 'LegacyGallery/2.0.0'.
For v2, it will simply the return the v2 endpoint itself

@yishaigalatzer @emgarten @pranavkm @feiling

### CODE REVIEW

**Comment by deepakaravindr** [comment]
Fixes https://github.com/NuGet/Home/issues/1185 and https://github.com/NuGet/Home/issues/1002 as well

File: src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Protocol.Core.Types
+{
+    public class ListCommandResource : INuGetResource
+    {
+        public string ListEndpoint { get; }
+        public ListCommandResource(string listEndpoint)
```
Comment (emgarten) [inline src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs:8]: add a new line above this

File: src/NuGet.Protocol.Core.v3/Constants.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@ public static class Types
 
     public static class ServiceTypes
     {
+        public const string VersionV2 = "/2.0.0";
```
Comment (emgarten) [inline src/NuGet.Protocol.Core.v3/Constants.cs:22]: these should be public static readonly

**Comment by emgarten** [comment]
:shipit:

File: src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            if (listEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(listEndpoint));
+            }
+
+            ListEndpoint = listEndpoint;
+        }
+
+        public Task<string> GetListEndpointAsync(CancellationToken token)
```
Comment (MeniZalzman) [inline src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs]: Same comment here on the Async pattern

File: src/NuGet.Protocol.Core.v2/ListCommandResourceV2Provider.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+            if (v2repo != null
+                && v2repo.V2Client != null
+                && !string.IsNullOrEmpty(v2repo.V2Client.Source))
+            {
+                // For a V2 package source, the source url is the list endpoint as well
+                listCommandResource = new ListCommandResource(v2repo.V2Client.Source);
+            }
+
+            return new Tuple<bool, INuGetResource>(listCommandResource != null, listCommandResource);
```
Comment (MeniZalzman) [inline src/NuGet.Protocol.Core.v2/ListCommandResourceV2Provider.cs:35]: Please split the return

File: src/NuGet.Protocol.Core.v3/ListCommandResourceV3Provider.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            if (serviceIndex != null)
+            {
+                var baseUrl = serviceIndex[ServiceTypes.LegacyGallery].FirstOrDefault();
+                if (baseUrl != null)
+                {
+                    listCommandResource = new ListCommandResource(baseUrl.AbsoluteUri);
+                }
+            }
+
+            return new Tuple<bool, INuGetResource>(listCommandResource != null, listCommandResource);
```
Comment (MeniZalzman) [inline src/NuGet.Protocol.Core.v3/ListCommandResourceV3Provider.cs:37]: Same here

**Comment by MeniZalzman** [comment]
:shipit:

File: src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs (hunk 1 above)
Comment (deepakaravindr) [inline src/NuGet.Protocol.Core.Types/Resources/ListCommandResource.cs:8]: Fixed

File: src/NuGet.Protocol.Core.v3/Constants.cs (hunk 2 above)
Comment (deepakaravindr) [inline src/NuGet.Protocol.Core.v3/Constants.cs:22]: Changed all the public const string in this static class to be static readonly

File: src/NuGet.Protocol.Core.v2/ListCommandResourceV2Provider.cs (hunk 4 above)
Comment (deepakaravindr) [inline src/NuGet.Protocol.Core.v2/ListCommandResourceV2Provider.cs:35]: Fixed

File: src/NuGet.Protocol.Core.v3/ListCommandResourceV3Provider.cs (hunk 5 above)
Comment (deepakaravindr) [inline src/NuGet.Protocol.Core.v3/ListCommandResourceV3Provider.cs:37]: Fixed

**Comment by deepakaravindr** [comment]
Merged with https://github.com/NuGet/NuGet3/commit/334a1cdcd269ed27412f0d70535357df570a2f72

**Pull request closed without merging**